### PR TITLE
Fixed: Mojo::DOM doesn't recognize end of comment (when it should) #2…

### PR DESCRIPTION
**Pull request by poti1**
…030.

https://github.com/mojolicious/mojo/issues/2030

### CODE REVIEW

File: t/mojo/dom.t [hunk 1]
```diff
@@ -2643,10 +2643,39 @@ subtest 'Comments' => sub {
...
-  is $dom->tree->[5][1], ' HTML4 ',             'right comment';
-  is $dom->tree->[7][1], ' bad idea -- HTML4 ', 'right comment';
+  is $dom->tree->[1][1], ' HTML5 ',                      'right comment';
+  is $dom->tree->[3][1], ' bad idea -- HTML5 ',          'right comment';
+  is $dom->tree->[5][1],  '<',                           'not support html style comments';
+  is $dom->tree->[6][1], "!-- HTML4 -- >\n",             'not support html style comments';
+  is $dom->tree->[7][1], '<',                            'not support html style comments';
+  is $dom->tree->[8][1], "!-- bad idea -- HTML4 -- >\n", 'not support html style comments';
+
+  # Issue #2030
```
Comment (kraih) [inline t/mojo/dom.t:2653]: We don't use comments like that.

File: lib/Mojo/DOM/HTML.pm [hunk 2]
```diff
@@ -29,15 +29,20 @@ my $TOKEN_RE = qr/
...
+        )
       |
         \[CDATA\[(.*?)\]\]                                                     # CDATA
       )
     |
       \?(.*?)\?                                                                # Processing Instruction
     |
       \s*((?:\/\s*)?[^<>\s\/0-9.\-][^<>\s\/]*\s*(?:(?:$ATTR_RE){0,32766})*+)   # Tag
-    )>
+    )>                                                               # Comment tag can end with a bang
```
Comment (kraih) [inline lib/Mojo/DOM/HTML.pm:45]: That comment tells me nothing.

File: lib/Mojo/DOM/HTML.pm [hunk 3]
```diff
@@ -29,15 +29,20 @@ my $TOKEN_RE = qr/
         (?:\s+\[.+?\])?                                                        # Int Subset
         \s*)
       |
-        --(.*?)--\s*                                                           # Comment
+        --(?|                                                                  # Comment
+            () -* !? (?=>)  # Empty comment
+            |
+            (.*?)
+            --!?(?=>)       # Comment end
```
Comment (kraih) [inline lib/Mojo/DOM/HTML.pm:36]: The regex formatting here is inconsistent with the surrounding regex.

File: t/mojo/dom.t [hunk 4]
```diff
@@ -2643,10 +2643,39 @@ subtest 'Comments' => sub {
 <!-- HTML4 -- >
 <!-- bad idea -- HTML4 -- >
 EOF
-  is $dom->tree->[1][1], ' HTML5 ',             'right comment';
-  is $dom->tree->[3][1], ' bad idea -- HTML5 ', 'right comment';
-  is $dom->tree->[5][1], ' HTML4 ',             'right comment';
-  is $dom->tree->[7][1], ' bad idea -- HTML4 ', 'right comment';
+  is $dom->tree->[1][1], ' HTML5 ',                      'right comment';
+  is $dom->tree->[3][1], ' bad idea -- HTML5 ',          'right comment';
+  is $dom->tree->[5][1],  '<',                           'not support html style comments';
```
Comment (kraih) [inline t/mojo/dom.t:2648]: That description is not proper english.

File: t/mojo/dom.t [hunk 5]
```diff
@@ -2643,10 +2643,39 @@ subtest 'Comments' => sub {
...
+    is $dom->tree->[2][0],    'text',     "$_: have text node";
+    is $dom->tree->[2][1],    ' ',        "$_: have 1st text string";
+    is $dom->tree->[3][0],    'tag',      "$_: have 1st tag node";
+    is $dom->tree->[3][1],    'p',        "$_: have tag string";
+    is $dom->tree->[3][4][0], 'text',     "$_: have sub text node";
+    is $dom->tree->[3][4][1], 'OK',       "$_: have sub text string";
+    is $dom->tree->[4][0],    'text',     "$_: have 2nd text node";
+    is $dom->tree->[4][1],    ' ',        "$_: have 2nd text string";
+    is $dom->tree->[5][0],    'comment',  "$_: have 2nd comment node";
+    is $dom->tree->[5][1],    ' ',        "$_: have 2nd comment string";
```
Comment (kraih) [inline t/mojo/dom.t:2672]: All these descriptions are also not proper english.

File: t/mojo/dom.t [hunk 6]
```diff
@@ -2643,10 +2643,39 @@ subtest 'Comments' => sub {
...
+  is $dom->tree->[8][1], "!-- bad idea -- HTML4 -- >\n", 'not support html style comments';
+
+  # Issue #2030
+  for (
+    '<!-->',
+    '<!--->',
+    '<!-- --!>',
+  ) {
+    my $dom = Mojo::DOM->new( "$_ <p>OK</p> <!-- -->" );
+    my $space = / / ? ' ' : '';
```
Comment (kraih) [inline t/mojo/dom.t:2660]: What is this supposed to do?

**Comment by kraih** [comment]
Please squash your commits. And split up individual bug fixes into separate PRs.

**Comment by kraih** [comment]
And for bug fixes where specs are relevant, please reference the according spec sections.

File: t/mojo/dom.t (hunk 1 above)
Comment (poti1) [reply t/mojo/dom.t:2653]: You mean: `# Issue #2030`?

No problem. I can remove it

**Comment by poti1** [comment]
> Please squash your commits. And split up individual bug fixes into separate PRs.

Should I close this PR and open another with a single squashed branch per PR?

(Just noticed you already did reply back)

**Comment by poti1** [comment]
> And for bug fixes where specs are relevant, please reference the according spec sections.

You mean that I should include a comment with a link to the specific spec?

**Comment by poti1** [comment]
Will open separate PRs

File: t/mojo/dom.t [hunk 7]
```diff
@@ -2643,10 +2643,33 @@ subtest 'Comments' => sub {
...
-  is $dom->tree->[5][1], ' HTML4 ',             'right comment';
-  is $dom->tree->[7][1], ' bad idea -- HTML4 ', 'right comment';
+  is $dom->tree->[1][1], ' HTML5 ',                      'right comment';
+  is $dom->tree->[3][1], ' bad idea -- HTML5 ',          'right comment';
+  is $dom->tree->[5][1], '<',                            'wrong comment';
+  is $dom->tree->[6][1], "!-- HTML4 -- >\n",             'wrong comment';
+  is $dom->tree->[7][1], '<',                            'wrong comment';
+  is $dom->tree->[8][1], "!-- bad idea -- HTML4 -- >\n", 'wrong comment';
+
+  for ('<!-->', '<!--->', '<!-- --!>') {
```
Comment (kraih) [inline t/mojo/dom.t:2653]: `$_` with such a wide scope is not good style.

Comment (kraih) [reply t/mojo/dom.t:2653]: In fact, i don't like loops in tests at all, they only make errors harder to find because line numbers get so much less useful.

**Pull request closed without merging**